engine/source_sym_engine.ml

```diff
@@ -0,0 +1,175 @@
+  let rec bprint_list ~sep bprint buf = function
+    | [] -> ()
+    | [x] -> bprint buf x
+    | x :: xs ->
+       bprintf buf "%a%t%a"
+         bprint x
+         sep
+         (bprint_list ~sep bprint) xs in
+  let break ntabs buf =
+    bprintf buf "\n%s" (BatList.init ntabs (fun _ -> "\t") |> String.concat "") in
+  let rec bprint_tree ntabs buf tree =
+    let sep = break (ntabs+1) in
+    let bprint_constructor buf k = match k with
+      | Variant s -> bprintf buf "Variant %s" s
+      | Int i -> bprintf buf "Int %d" i
+      | Bool b -> bprintf buf "Bool %b" b
+      | String s -> bprintf buf "String \"%s\"" s
+      | Tuple narity -> bprintf buf "Tuple[%d]" narity
+      | Nil ->  bprintf buf "Nil"
+      | Cons -> bprintf buf "Cons"
+    in
+    match tree with
+    | Empty -> bprintf buf "Empty"
+    | Leaf expr ->
+      bprintf buf
+        "Leaf='%a'"
+        bprint_source_expr expr
+    | Node (k_cst_list, fallback_cst) ->
+      bprintf buf "Node:{\
+                   %a \
+                   %t} Fallback: %a"
+        (bprint_list ~sep:sep
+           (fun buf (k,cst) -> bprintf buf "%t%a -> %t%a"
+               sep
+               bprint_constructor k
+               (break (ntabs+2))
+               (bprint_tree (ntabs+1)) cst))
+        k_cst_list
+        (break ntabs)
+        (bprint_tree (ntabs+1)) fallback_cst
+  in
+  let buf = Buffer.create 42 in
+  bprint_tree 0 buf stree;
+  BatIO.write_line BatIO.stdout (Buffer.contents buf)
+
+type row = pattern list * source_expr
+
+type hashable_constructor = | HConstructor of constructor | HWildcard
+
+let sym_exec source =
+  let group_constructors rows : (constructor * row list) list * row list =
+    let grouptbl = Hashtbl.create 42
+    in
+    let (ksttbl: (source_blackbox , int) Hashtbl.t) = Hashtbl.create 42
+    in
+    let narity_of_k = function
+      | Nil -> 0
+      | Int _ | Bool _ | String _ -> 1
+      | Cons -> 2
+      | Tuple n -> n
+      | Variant v -> Hashtbl.find ksttbl v
+    in
+    let nary_wildcard n wildcards =
+      let rec inc_narity n ((wlist, expr):pattern list * source_expr) =
+        assert ( n >= 0);
+        if n > 0
+        then inc_narity (n-1) (Wildcard::wlist, expr)
+        else (wlist, expr)
+      in
+      List.map (inc_narity (n-List.length wildcards)) wildcards
+    in
+    let rec first_scan : pattern list -> unit = function
+      | [] -> ()
+      | (pattern::ptl) ->
+        match pattern with
+        | Wildcard -> ()
+        | Or (p1, p2) -> first_scan [p1; p2]
+        | As (p, _) -> first_scan [p]
+        | Constructor (k, plist) ->
+          match k with
+          | Variant v -> 
+            let len_ptl = List.length plist in
+            begin match Hashtbl.find_opt ksttbl v with 
+              | Some (narity:int) -> if len_ptl > narity then Hashtbl.replace ksttbl v len_ptl
+              | None -> Hashtbl.add ksttbl v len_ptl
+            end;
```

Comment:
I think that two behaviors make sense here:
- either you assume that all occurrences of a variant constructor have the exact same arity,
  and just use `Hashtbl.replace ksttbl v len_ptl` without checking that a binding exists
- or you enforce that, if a binding already exists, it must be the exact same arity or you fail.

The current code computes a maximum, but that is not correct in incompatible-arities scenario (I don't see what one could do correctly in those scenarios).